cranelift/codegen/src/isa/x86/abi.rs

```diff
@@ -799,38 +826,27 @@ fn insert_common_prologue(
         }
     }
 
-    // Now that RSP is prepared for the function, we can use stack slots:
+    // With the stack pointer adjusted, save any callee-saved floating point registers via offset
+    // FPR saves are at the highest addresses of the local frame allocation, immediately following the GPR pushes
     let mut last_fpr_save = None;
-    if let Some(fpr_slot) = fpr_slot {
-        debug_assert!(csrs.iter(FPR).len() != 0);
-
-        // `stack_store` is not directly encodable in x86_64 at the moment, so we'll need a base
-        // address. We are well after postopt could run, so load the CSR region base once here,
-        // instead of hoping that the addr/store will be combined later.
-        // See also: https://github.com/bytecodealliance/wasmtime/pull/1198
-        let stack_addr = pos.ins().stack_addr(types::I64, *fpr_slot, 0);
-
-        // Use r11 as fastcall allows it to be clobbered, and it won't have a meaningful value at
-        // function entry.
-        pos.func.locations[stack_addr] = ir::ValueLoc::Reg(RU::r11 as u16);
-
-        let mut fpr_offset = 0;
 
-        for reg in csrs.iter(FPR) {
-            // Append param to entry Block
-            let csr_arg = pos.func.dfg.append_block_param(block, types::F64X2);
+    for (i, reg) in csrs.iter(FPR).enumerate() {
+        // Append param to entry block
+        let csr_arg = pos.func.dfg.append_block_param(block, types::F64X2);
 
-            // Since regalloc has already run, we must assign a location.
-            pos.func.locations[csr_arg] = ir::ValueLoc::Reg(reg);
+        // Since regalloc has already run, we must assign a location.
+        pos.func.locations[csr_arg] = ir::ValueLoc::Reg(reg);
 
-            last_fpr_save =
-                Some(
-                    pos.ins()
-                        .store(ir::MemFlags::trusted(), csr_arg, stack_addr, fpr_offset),
-                );
+        // Offset to where the register is saved relative to RSP, accounting for FPR save alignment
+        let offset = ((i + 1) * types::F64X2.bytes() as usize) as i64
+            + (stack_size & isa.pointer_bytes() as i64);
```

Comment:
since `local_stack_size` ought to be 16-byte aligned, shouldn't `stack_size & isa.pointer_bytes()` always be zero? It seems like we could lift this up to an assert on function entry, if so.

Reply:
`stack_size` (i.e. the adjustment to RSP) is not 16-byte aligned as it might be aligning RSP back to a 16-byte alignment.

Take for instance a prologue in a non-leaf function that pushes the frame pointer and then saves a single GPR.  The push of the frame pointer realigned the stack from the push of return address, but the push of the GPR causes a misalignment.  The local stack allocation will be +0x8 to account for that, which means `stack_size` itself is not a multiple of 16, but 8.

This lines cause causes the offset to be +0x8 when `stack_size` itself is +0x8 so that it gets subtracted out from the final offset from the post-adjusted RSP.  The end result is that the offset should always be a multiple of 16 from a post-adjusted 16-byte aligned RSP.

Reply:
That said, I'm totally open to calculating this in a clearer way if you have any suggestions!

Reply:
That example helps, thanks. How's `% types::F64X2.bytes()` instead of `& isa.pointer_bytes()` sound to you? That picks an aligned offset for the type to store, ought to work for `isa.pointer_bytes() == 4`, and makes extension for ymm preservation (if we get to that in the future) clearer (they don't _need_ 32-byte alignment, but iiuc are somewhat penalized for misaligned access)

Reply:
Works for me.  I'll get that changed!